queue.c

```diff
@@ -2491,19 +2518,169 @@ static void prvCopyDataFromQueue( Queue_t * const pxQueue,
+                    if( listLIST_IS_EMPTY( &( pxQueue->xTasksWaitingToReceive ) ) == pdFALSE )
+                    {
+                        if( xTaskRemoveFromEventList( &( pxQueue->xTasksWaitingToReceive ) ) != pdFALSE )
+                        {
+                            /* The task waiting has a higher priority so record that
+                             * a context switch is required. */
+                            vTaskMissedYield();
+                        }
+                        else
+                        {
+                            mtCOVERAGE_TEST_MARKER();
+                        }
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+                #endif /* configUSE_QUEUE_SETS */
+
+                --cTxLock;
+            }
+
+            pxQueue->cTxLock = queueUNLOCKED;
+        }
+        queueEXIT_CRITICAL( pxQueue );
+
+        /* Do the same for the Rx lock. */
+        queueENTER_CRITICAL( pxQueue );
+        {
+            int8_t cRxLock = pxQueue->cRxLock;
+
+            while( cRxLock > queueLOCKED_UNMODIFIED )
+            {
+                if( listLIST_IS_EMPTY( &( pxQueue->xTasksWaitingToSend ) ) == pdFALSE )
+                {
+                    if( xTaskRemoveFromEventList( &( pxQueue->xTasksWaitingToSend ) ) != pdFALSE )
+                    {
+                        vTaskMissedYield();
+                    }
+                    else
+                    {
+                        mtCOVERAGE_TEST_MARKER();
+                    }
+
+                    --cRxLock;
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            pxQueue->cRxLock = queueUNLOCKED;
+        }
+        queueEXIT_CRITICAL( pxQueue );
+    }
+#endif /* #if ( portUSING_GRANULAR_LOCKS == 0 ) */
+/*-----------------------------------------------------------*/
+
+#if ( ( portUSING_GRANULAR_LOCKS == 1 ) && ( configNUMBER_OF_CORES > 1 ) )
+    static void prvLockQueueForTasks( Queue_t * const pxQueue )
+    {
+        /* Disable preemption so that the current task cannot be preempted by another task */
+        vTaskPreemptionDisable( NULL );
+
+        /* Lock the queue data group so that we can suspend the queue atomically */
+        queueENTER_CRITICAL( pxQueue );
+
+        /* Suspend writing and reading to/from the queue */
+        if( pxQueue->cRxLock == queueUNLOCKED )
+        {
+            pxQueue->cRxLock = queueLOCKED_UNMODIFIED;
+        }
+
+        if( pxQueue->cTxLock == queueUNLOCKED )
+        {
+            pxQueue->cTxLock = queueLOCKED_UNMODIFIED;
+        }
+
+        /* Keep holding xTaskSpinlock after unlocking the data group to prevent tasks
+         * on other cores from accessing the queue while it is suspended. */
+        portGET_SPINLOCK( portGET_CORE_ID(), &( pxQueue->xTaskSpinlock ) );
+
+        queueEXIT_CRITICAL( pxQueue );
```

Comment:
Can we adapt the following change to acquire task spinlock with interrupt enabled?
```suggestion
        queueEXIT_CRITICAL( pxQueue );

        /* Keep holding xTaskSpinlock after unlocking the data group to prevent tasks
         * on other cores from accessing the queue while it is suspended. */
        portGET_SPINLOCK( portGET_CORE_ID(), &( pxQueue->xTaskSpinlock ) );
```

Reply:
Made the suggested change as `portGET_SPINLOCK()` is thread safe.